ext/phar/dirstream.c

```diff
@@ -185,64 +166,46 @@ static php_stream *phar_make_dirstream(char *dir, HashTable *manifest) /* {{{ */
 		return php_stream_alloc(&phar_dir_ops, data, NULL, "r");
 	}
 
-	zend_hash_internal_pointer_reset(manifest);
-
-	while (FAILURE != zend_hash_has_more_elements(manifest)) {
-		if (HASH_KEY_NON_EXISTENT == zend_hash_get_current_key(manifest, &str_key, &unused)) {
-			break;
-		}
+	zend_string *str_key;
+	ZEND_HASH_FOREACH_STR_KEY(manifest, str_key) {
+		ZEND_ASSERT(str_key);
 
-		keylen = ZSTR_LEN(str_key);
+		size_t keylen = ZSTR_LEN(str_key);
 		if (keylen <= dirlen) {
 			if (keylen == 0 || keylen < dirlen || !strncmp(ZSTR_VAL(str_key), dir, dirlen)) {
-				if (SUCCESS != zend_hash_move_forward(manifest)) {
-					break;
-				}
 				continue;
 			}
 		}
 
 		if (*dir == '/') {
 			/* root directory */
-			if (keylen >= sizeof(".phar")-1 && !memcmp(ZSTR_VAL(str_key), ".phar", sizeof(".phar")-1)) {
+			if (zend_string_starts_with_literal(str_key, ".phar")) {
 				/* do not add any magic entries to this directory */
-				if (SUCCESS != zend_hash_move_forward(manifest)) {
-					break;
-				}
 				continue;
 			}
 
-			if (NULL != (found = (char *) memchr(ZSTR_VAL(str_key), '/', keylen))) {
+			const char *has_slash = (char *) memchr(ZSTR_VAL(str_key), '/', keylen);
+			if (has_slash) {
 				/* the entry has a path separator and is a subdirectory */
-				entry = (char *) safe_emalloc(found - ZSTR_VAL(str_key), 1, 1);
-				memcpy(entry, ZSTR_VAL(str_key), found - ZSTR_VAL(str_key));
-				keylen = found - ZSTR_VAL(str_key);
-				entry[keylen] = '\0';
-			} else {
-				entry = (char *) safe_emalloc(keylen, 1, 1);
-				memcpy(entry, ZSTR_VAL(str_key), keylen);
-				entry[keylen] = '\0';
+				keylen = has_slash - ZSTR_VAL(str_key);
 			}
+			entry = (char *) safe_emalloc(keylen, 1, 1);
```

Comment:
You don't need to cast this to `char *`.

Reply:
This was copied and pasted from the original code but will remove.